[fix] Fix VPN-client template switch bug with same VPN server

shwetd19 opened the pull request:
Resolves a bug where VPN configuration variables become unresolved when switching between VPN templates using the same VPN server

## Checklist

- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [x] I have manually tested the changes proposed in this pull request.
- [x] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- N/A I have updated the documentation.

## Reference to Existing Issue

Closes #973.

## Description of Changes

This PR fixes an issue in the VPN client management code that occurs when switching between VPN templates that use the same VPN server. The bug popups when simultaneously removing one VPN template while adding another that uses the same VPN server.

### The Problem

The issue was occuring due to how VPN clients are managed during template changes:

1. When adding a new template (`post_add`):
   - Creating a new VPN client for the new template
   - Deletes existing VPN clients not in the current template list
   - This could delete VPN clients still needed by other templates

2. When removing a template (`post_remove`):
   - Deletes VPN clients associated with the removed template
   - This could delete VPN clients still needed by remaining templates

As a result, the configuration could end up with no VPN client object, causing VPN configuration variables to remain unresolved.

### The Solution

The fix modifies the `manage_vpn_clients` method to:

1. Tracking all of current VPN templates and their associated VPNs upfront
2. For `post_add`:
   - Create new VPN clients first
   - Only delete VPN clients whose VPNs are not associated with any current template
3. For `post_remove`:
   - Only delete VPN clients if their VPN is not used by any remaining template

This soln ensures VPN clients are preserved when switching between templates that use the same VPN server, maintaining proper configuration variable resolution.

pandafy reviewed (changes requested):
@shwetd19 kindly add a failing test for the bug described in the issue. A fix for this issue should not remove the existing features, i.e. the existing tests should not fail due to missing logic.

shwetd19 commented on openwisp_controller/config/base/config.py:
```diff
@@ -245,7 +245,6 @@ def manage_vpn_clients(cls, action, instance, pk_set, **kwargs):
         instance is using templates which have type set to "VPN"
         and "auto_cert" set to True.
         This method is called from a django signal (m2m_changed)
-        see config.apps.ConfigConfig.connect_signals
```
@pandafy Sorry for removing the comments - that was unintentional. I've restored the original docstring comments in my latest commit

shwetd19 commented on openwisp_controller/config/base/config.py:
```diff
@@ -256,46 +255,47 @@ def manage_vpn_clients(cls, action, instance, pk_set, **kwargs):
+            # Only delete VPN clients if their VPN is not used by any remaining template
+            for template in templates.filter(type='vpn'):
+                if template.vpn_id not in current_vpns:
+                    instance.vpnclient_set.filter(vpn=template.vpn).delete()
```
@pandafy Thanks for catching this! I understand the concern about potential duplicate VPN clients. I've modified the code to use your suggested approach:

1. In post_add:
   - First create new VPN clients 
   - Then cleanup any unused clients with exclude(vpn__in=current_vpns)

2. In post_remove:
   - Simply cleanup all unused clients with exclude(vpn__in=current_vpns)

This ensures we won't have duplicate VPN clients for the same config object while preserving the necessary clients. The full cleanup at the end handles any edge cases where duplicate clients might have been created.

pandafy replied:
We need a test which verifies this.

shwetd19 commented:
Hey @pandafy I've implemented the suggested changes, can you pls check once

shwetd19 commented:
Hey @pandafy @nemesifier can you please reivew this PR ?
pls let me know if any changes are required, Thanks!

pandafy reviewed (changes requested):
@shwetd19 add a test. You can take reference from this existing test here 

https://github.com/openwisp/openwisp-controller/blob/129a42f957166fe6652e5181e9bf03eadc2c3984/openwisp_controller/config/tests/test_admin.py#L1623-L1704

The test should do the following operations: 
1. Create a VPN object 
2. Create two templates with the same vpn server. (say template1 and template 2)
```python
        vpn = self._create_vpn()
        template1 = self._create_template(
            name='vpn-test-1',
            type='vpn',
            vpn=vpn,
            config={},
            auto_cert=True,
            default=True, # This will auto-apply template1 to the device
        )
        template2 = self._create_template(
            name='vpn-test-2',
            type='vpn',
            vpn=vpn,
            config={},
            auto_cert=True,
        )
```
3. Using the Django admin, apply tmplate2 and remove template1 from the device 
4. Verify there's only 1 VpnClient present for the config
5. Verify that the variables are correctly resolved in the config using the following: 
```python
       self.assertEqual(
            config.backend_instance.config['openvpn'][0]['cert'],
            f'/etc/x509/client-{vpn.pk.hex}.pem',
        )
````

We cannot accept a bugfix patch with a test. Ensure that the test fails without the changes you've made in openwisp_controller/config/base/config.py

nemesifier reviewed (changes requested):
@shwetd19 as @pandafy indicated, we need a tests which replicates the issue and fails without the patch. Please let us know if you have doubts on how to write it.

nemesifier reviewed (commented):
@shwetd19 

> Just to confirm, should the test also include a check for the behavior when switching back and forth between `template1` and `template2` multiple times? Or is it sufficient to test a single switch from `template1` to `template2`?
> 
> I’ll proceed with writing the test and will update the PR soon. If there’s anything else I should keep in mind, please let me know.

Switching back would from template2 to template1 would be nice and shouldn't involve a lot more effort.

> Thanks again for your guidance!

Welcome!

PS: look at the existing tests to get inspiration on how to write your test, try to be consistent with the rest of the codebase please.

shwetd19 commented:
Hey @pandafy @nemesifier 

I've added the fix here as : Added a test to replicate the bug and verify the fix, which creates two VPN templates using the same VPN server, switches between them, and verifies that only one `VpnClient` object exists for the configuration and that VPN configuration variables are resolved correctly, ensuring the test fails without the fix and passes with it applied.



also I've tested it locally as here :
![image](https://github.com/user-attachments/assets/9f4126d9-1b5f-4966-958e-a5301b44e1d6)

pandafy commented on openwisp_controller/config/base/config.py:
```diff
@@ -812,5 +812,63 @@ def manage_backend_changed(cls, instance_id, old_backend, backend, **kwargs):
+        # Create a device and assign template1
+        device = self._create_device()
+        config = device.config
+        config.templates.add(template1)
+        config.save()
```
We have used `default=True` while creating the template, thus we don't need to add the template1 to the Config object.

pandafy commented on openwisp_controller/config/base/config.py:
```diff
@@ -812,5 +812,63 @@ def manage_backend_changed(cls, instance_id, old_backend, backend, **kwargs):
             old_templates = device_group.templates.filter(backend=old_backend)
         config.manage_group_templates(templates, old_templates, not created)
 
+    def test_vpn_template_switch(self):
```
@shwetd19 why did you add this test in the model definition.

pandafy reviewed (changes requested):
@shwetd19 I don't see any progress on this PR after my last review. 

Your recent commits have deleted the changes to the vpn_templates logic, kindly rectify it.

pandafy commented:
> also I've tested it locally as here : ![image](https://private-user-images.githubusercontent.com/119885670/413010449-9f4126d9-1b5f-4966-958e-a5301b44e1d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk1Mjk3NzEsIm5iZiI6MTczOTUyOTQ3MSwicGF0aCI6Ii8xMTk4ODU2NzAvNDEzMDEwNDQ5LTlmNDEyNmQ5LTFiNWYtNDk2Ni05NThlLWE1MzAxYjQ0ZTFkNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxNFQxMDM3NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZWI3ZDNlOWY0YzQzNWFjYTIxOGVjMjkzMTk0MzExYWJiZjJjY2U0YjdkNzM3NTc2MWJhNGIzYmRlOGUwN2NiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bQQV7iMEVo9frR4blWQrGs0oRhPp7kgEaKkEfXO5UmU)

You ran the QA checks, not test suite. Read the docs on how to run the tests.

nemesifier commented on openwisp_controller/config/base/config.py:
```diff
@@ -256,7 +256,7 @@ def manage_vpn_clients(cls, action, instance, pk_set, **kwargs):
         if action == 'post_clear':
             if instance.is_deactivating_or_deactivated():
-                # If the device is deactivated or in the process of deactivating, then
+                # If the device is deactivated or in the process of deactivatiing, then
```
```suggestion
                # If the device is deactivated or in the process of deactivating, then
```

nemesifier reviewed (commented):
@shwetd19 have you verified that `test_vpn_template_switch` fails with a clear error message without the code fix?

nemesifier reviewed (changes requested):
CI build still failing @shwetd19.

nemesifier reviewed (approved):
I did some local manual testing to confirm the solution works and polished the new automated test.

Great work @shwetd19 @pandafy :clap:
